src/crimson/os/seastore/transaction.h

```diff
@@ -343,6 +343,18 @@ class Transaction {
     return ret;
   }
 
+  const auto &get_pre_alloc_list() {
+    return pre_alloc_list;
+  }
+
+  bool is_issued_ool_writes() {
+    auto& stats = get_ool_write_stats();
```

Comment:
I think relying on `get_ool_write_stats()` isn't specific -- the issue is only related to `RandomBlockOolWriter`.

Continue the discussion of https://github.com/ceph/ceph/pull/59720#discussion_r1756219827, how about introducing a shared ptr of:
```c++
struct rbm_pending_ool_t {
  bool is_conflicted = false;
  std::list<CachedExtentRef> pending_extents;
};
```

So that both `Transaction` and `RandomBlockOolWriter::do_write()` can hold the same `rbm_pending_ool_t` independently, and so that:
* Upon `RandomBlockOolWriter::do_write()`, add the affecting extents to `rbm_pending_ool_t::pending_extents`.
* Upon transaction conflict, we can set `rbm_pending_ool_t::is_conflicted` through `Transaction` to true to notify `RandomBlockOolWriter::do_write()`.
* When `RandomBlockOolWriter::do_write()` finishes write, it can refer to `rbm_pending_ool_t::is_conflicted` to determine whether need to cancel the reservation, even if the `Transaction` is already destructed/reset/retrying.

This is a rough idea to decouple transaction reset and RandomBlockOolWriter post ool writes, so there is no assumption about which finishes/destructs first. How do you think?

Reply:
Sounds like a better approach. I've updated the commit. Can you take a look?